Windows: A module for creating Toast notifications on Modern Windows versions.

jhawkesworth opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - New Module Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
win_toast
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (win_toast 924e87f6b7) last updated 2017/07/12 08:24:37 (GMT +100)
  config file = None
  configured module search path = [u'/home/jon/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jon/ansible-dev/lib/ansible
  executable location = /home/jon/ansible-dev/bin/ansible
  python version = 2.7.6 (default, Jun 22 2015, 17:58:13) [GCC 4.8.2]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
So I thought it would be nice to be able to send modern UWP style notifications to newer windows hosts, as an alternative to win_msg.

Since this only works on Windows 10 and Server 2016 I put a windows version check in.

Testing seems to have shown that the notification service is only available when there is a logged in user, further reducing the utility of this module. :-(

Currently a work in progress as I have not found a way to make the notifications persist in Action Center.
Without this the notifications disappear after a couple of seconds, making them so ephemeral as to be useless.

Also setting popup to false does not have the intended effect of making the notification only appear in Action Center - the module runs but doesn't appear to do anything.

Likely would want to pick some better defaults for the group and tag.
Maybe also select the toast template to use (see https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Notifications.ToastTemplateType) , and add a way to specify the image to use.


<!--- Paste verbatim command output below, e.g. before and after your change -->
```
ansible 10 -m win_toast -a "expire_mins=17 msg='Hi, it looks like you were doing something\nimportant, but now I have interupted you.\nAm I clever, or what?' title='Notification from Ansible' popup=False"
TENSY | SUCCESS => {
    "changed": false,
    "expire_at": "12/07/2017 08:35:24",
    "expire_at_utc": "12 July 2017 07:35:24",
    "expire_mins": 17,
    "failed": false,
    "group": "Powershell",
    "msg": "Hi, it looks like you were doing something\nimportant, but now I have interupted you.\nAm I clever, or what?",
    "popup": false,
    "runtime_seconds": 0.1881824,
    "sent_localtime": "12 July 2017 08:18:24",
    "tag": "Ansible",
    "title": "Notification from Ansible"
}
```

jhawkesworth commented:
Hmm, likely the problem is that once powershell command completes there's no app left for action center to communicate with.  If that's the case probably not worth the trouble.
Given this example here https://msdn.microsoft.com/library/1d20ed75-e24a-4e60-91ab-cfcbe902a68e I'm wondering if in powershell the appid is inherent so as soon as the module code exits, its game over for the notification.

dagwieers commented:
+label new_module

jhawkesworth commented:
I've confirmed that, as it stands you need powershell to continue to be running for the notification to hang around (by adding 
start-sleep -Seconds ($expire_mins * 60) 
after the notification is displayed.

You could I suppose run as an async task - but that's ugly and inconsistent with the other windows message modules.   Or create a scheduled task to run and display the notification.  Or create an emphemeral service.  All of which is much code just to display a notification.  Will continue to think about this but I'm open to suggestions fro a way to solve this.

mikeeaton83 commented:
I'm not sure that it's completely useless.. Jon. One quick example of where this would be useful, at least in my situation is to alert any logged in users that Ansible is currently configuring the server they're on and to close any unsaved work immediatly - in case of a reboot. All Windows servers I create now are done with Ansible and are Windows Server 2016.

Michael

jhawkesworth commented:
True, but specifically for reboots, win_reboot can already send a custom message and wait a bit before actually doing the reboot.

http://docs.ansible.com/ansible/latest/win_reboot_module.html

I like the idea of toast for 'in advance' type warnings, the day before kind of thing.  Just need to work out an acceptable way to leave a process running for the notification to belong to, which doesn't leave something hanging around on the windows machine indefinitely.

dagwieers commented:
So the module freeze deadline for v2.4 is getting closer (2017-08-29). Now is the time to finish up, get it reviewed and merged with no delay.

alikins commented:
Something is wrong with that branch, as this is showing 1234 modified files, making tools like https://ansible.sivel.net/byfile.html confused.

jhawkesworth commented:
sorry I have clearly messed up this branch with an accidental pull - please ignore I will sort it out as soon as I can.

jhawkesworth commented:
Ugh.  Branch still hosed. Will attempt more git-fu.

Commit:
replace duff commit version of win_toast

jhawkesworth commented:
ugh sorry for the ping noise (again).

Commit:
change expire_mins to expire_secs and add example showing use of async

jhawkesworth commented:
ready_for_review

Still not as slick as you might like but now at least usable.  By adding a sleep statement and running via async, you can at least fire and forget the notification and it will persist until async timeout.

Commit:
fix metadata version to keep sanity --test validate-modules happy

jhawkesworth commented:
doh, how did I miss metadata version changing!
I have a fix for that, will add a basic integration test too.

dagwieers commented:
We may still get this in for v2.4 ? :-)

dagwieers commented on lib/ansible/modules/windows/win_toast.ps1:
```diff
@@ -0,0 +1,101 @@
+#
+# Copyright 2017, Jon Hawkesworth (@jhawkesworth) <figs@unity.demon.co.uk>
+#
+# Ansible is free software: you can redistribute it and/or modify
```
We now have a one-line GPL header for files.

jhawkesworth replied:
Thanks will fix that

dagwieers commented on lib/ansible/modules/windows/win_toast.ps1:
```diff
@@ -0,0 +1,101 @@
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+
+# $expire_at = Get-AnsibleParam -obj $params -name "msg" -type "str"
+$expire_secs = Get-AnsibleParam -obj $params -name "expire_secs" -type "int" -default 45
```
I am not fond of parameter options with the unit in the parameter name. We have had that discussion before, and actually decided to raise this to the core team. But usually the unit is part of the parameter description.

Besides, other parameters are using `_sec`, this is the first time `_secs` is proposed.

jhawkesworth replied:
I've actually changed to 'expire_seconds' (same I as used in win_msg) so it won't be _secs (which I think is not good - whole words are always clearer).

dagwieers replied:
Noted and updated the Core agenda ticket.

dagwieers commented on lib/ansible/modules/windows/win_toast.ps1:
```diff
@@ -0,0 +1,101 @@
+    expire_at = $expire_at.ToString()
+    expire_at_utc = $expire_at_utc
+    group = $group    
+    msg = $msg
```
So I have asked myself this in the past as well, because I see various modules putting the input parameters also as result values. What is the purpose of this practice ? I don't see any, the input was provided by the user, so it should be the same as the result. Why would we register that stuff again ? Who would ever need it coming from the module.

In my opinion we should stop doing this, and return back to minimal return values that have a use (i.e. new output).

Also, this is already part of the invocation output anyway if you provide a higher verbosity.

jhawkesworth replied:
I suppose I do it because if the values are calculated as part of module params then you can use the calculated version from the registered results.

A contrived (and untested) example
```yaml

win_toast:
  msg: "Warning {{release_name}} starts at {{ scheduled_release_time }}"
register: toast_message

# warn again
win_toast:
   msg: toast_message.msg
```

dagwieers replied:
I understand, but you already have the content of that message available to you. One could define it elsewhere (vars, set_fact) and use that variable twice.

dagwieers commented on lib/ansible/modules/windows/win_toast.py:
```diff
@@ -0,0 +1,126 @@
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
```
Same here, use the one-line GPL header.

jhawkesworth replied:
will do

dagwieers commented on lib/ansible/modules/windows/win_toast.py:
```diff
@@ -0,0 +1,126 @@
+    description:
+      - If false, the notification will not pop up and will only appear in the Action Center.
+    type: bool
+    default: 'Yes'
```
Lowercase 'yes' please :-)

jhawkesworth replied:
will fix

dagwieers commented:
Only 2 real issues for me, both related to the user interface. I will put both also on the agenda of today's meeting. And both issues are not blocking for me, it would be nice to have general consensus either way.

jhawkesworth commented:
Thanks for the review @dagwieers 
Will try to push fix and integration test at lunchtime.

Commit:
code review fixes and change expire_secs to expire_seconds

Commit:
add first pass integration tests for win_toast

dagwieers reviewed (approved):
LGTM

Commit:
win_toast no longer fails if there are no logged in users to notify (it sets a toast_sent false if this happens)

Commit:
yaml lint clean up of setup.yml in win_toast integration tests

dagwieers commented on lib/ansible/modules/windows/win_toast.ps1:
```diff
@@ -0,0 +1,97 @@
+     }
+  } catch {
+    $excep= $_
+    Fail-Json $result "Failed to create toast notifier, exception was $($excep.Exception.Message) stack trace $($excep.ScriptStackStrace)".
```
A small one I just noticed, not a deal breaker. We now tend to do
```powershell
   Fail-Json -obj $result -message "Failed to create toast notifier: $($excep.Exception.Message) with stack trace $($excep.ScriptStackStrace)".
```
For bonus points you can add the stack trace to the standard result value `exception` meand for stack traces:
```powershell
   $result.exception = $excep.ScriptStackStrace
   Fail-Json -obj $result -message "Failed to create toast notifier: $($excep.Exception.Message)".
```
PS Yes, bad name, it should have been `traceback` or something similar. See: http://docs.ansible.com/ansible/latest/common_return_values.html#exception

Commit:
improve exception and stack trace if the notifier cannot be created, following feedback from dag

jhawkesworth commented:
@dagwieers - I fixed up the Fail-Json to return the stack trace and message as requested.

Commit:
removed unwanted 'echo' input parameters from return vals; added to CHANGELOG.md, removed _seconds units from module params; updated tests to match

jhawkesworth commented:
Removed 'input' parameters from output.
Removed _seconds from 'expire_seconds' module param
Renamed 'runtime_seconds' to 'time_taken' in module return vals.
added to CHANGELOG.md (perhaps optimistically)

jborean93 commented on lib/ansible/modules/windows/win_toast.ps1:
```diff
@@ -0,0 +1,91 @@
+$osversion = [Environment]::OSVersion
+$lowest_version = 10
+if ($osversion.Version.Major -lt $lowest_version ) {
+   Fail-Json $result "Sorry, this version of windows, $osversion, does not support Toast notifications.  Toast notifications are available from version $lowest_version" 
```
Does this work when `$result` isn't defined yet?

Edit: Looks like it does.

jborean93 commented:
Works for me, thanks @jhawkesworth
